### PR TITLE
push filter & pull validator

### PR DESCRIPTION
**Pull request by Sandychuang8**
Adding push filter and pull validator

### CODE REVIEW

File: src/Couchbase.Lite.Shared/API/Sync/Replicator.cs [hunk 1]
```diff
@@ -262,10 +262,40 @@ private static void StatusChangedCallback(C4Replicator* repl, C4ReplicatorStatus
...
 
+        [MonoPInvokeCallback(typeof(C4ReplicatorPushFilterFunction))]
+        private static bool PushFilterCallback(FLSlice docID, FLDict* dict, void* context)
+        { 
+            var replicator = GCHandle.FromIntPtr((IntPtr)context).Target as Replicator;
+            var docIDStr = docID.CreateString();
+            return replicator.PushFilterCallback(docIDStr, dict);
+        }
+
+        bool PushFilterCallback(string docID, FLDict* dict)
```
Comment (borrrden) [inline src/Couchbase.Lite.Shared/API/Sync/Replicator.cs]: Add "private" explicitly

File: src/Couchbase.Lite.Shared/API/Sync/Replicator.cs [hunk 2]
```diff
@@ -262,10 +262,40 @@ private static void StatusChangedCallback(C4Replicator* repl, C4ReplicatorStatus
...
+            var replicator = GCHandle.FromIntPtr((IntPtr)context).Target as Replicator;
+            var docIDStr = docID.CreateString();
+            return replicator.PullValidateCallback(docIDStr, dict);
+        }
+
+        bool PullValidateCallback(string docID, FLDict* dict)
+        {
+            if (Config.PullFilter == null)
+                return true;
+            var doc = Config.Database.GetDocument(docID);
```
Comment (borrrden) [inline src/Couchbase.Lite.Shared/API/Sync/Replicator.cs]: I'm not sure this is the way that was intended.  What is coming with "dict"?  If it is the document properties then those should be used instead.  Otherwise there will be a database call for every item in the pull replication which will slow things down a lot (ditto for the push callback).

File: src/LiteCore/src/LiteCore.Shared/Interop/C4Replicator.cs [hunk 3]
```diff
@@ -32,35 +32,38 @@ namespace LiteCore.Interop
...
+    [return: MarshalAs(UnmanagedType.U1)]
+    internal unsafe delegate bool C4ReplicatorPushFilterFunction(FLSlice docID, FLDict* body,
+        void* context);
 
     [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
-    internal unsafe delegate void C4ReplicatorPushFilterFunction(FLSlice docID,
-        FLDict* body, void* context);
+    [return: MarshalAs(UnmanagedType.U1)]
+    internal unsafe delegate bool C4ReplicatorValidationFunction(FLSlice docID, FLDict* body, 
+        void* context);
```
Comment (borrrden) [inline src/LiteCore/src/LiteCore.Shared/Interop/C4Replicator.cs]: These probably don't require different delegates since the signatures are the same.  They are both "Validation functions."

File: src/LiteCore/src/LiteCore.Shared/Interop/C4Replicator.cs [hunk 4]
```diff
@@ -61,9 +56,9 @@ internal sealed class ReplicatorParameters : IDisposable
         private C4ReplicatorParameters _c4Params;
         private C4ReplicatorBlobProgressCallback _onBlobProgressUpdated;
         private C4ReplicatorStatusChangedCallback _onStatusChanged;
-        private C4ReplicatorDocumentEndedCallback _onDocumentEnded;
-        private C4ReplicatorPushFilterFunction _pushFilter;
-        private C4ReplicatorValidationFunction _validation;
+        private C4ReplicatorDocumentErrorCallback _onDocumentEnded;
```
Comment (borrrden) [inline src/LiteCore/src/LiteCore.Shared/Interop/C4Replicator.cs]: Why does this keep changing back and forth?

Comment (Sandychuang8) [reply src/LiteCore/src/LiteCore.Shared/Interop/C4Replicator.cs]: I modified C4ReplicatorDocumentErrorCallback to C4ReplicatorDocumentEndedCallback in feature replication eventing. But this PR should not have that change yet...so I changed it back..

File: src/Couchbase.Lite.Shared/API/Sync/Replicator.cs (hunk 2 above)
Comment (Sandychuang8) [reply src/Couchbase.Lite.Shared/API/Sync/Replicator.cs]: I actually don't know how to get info out FLDict.

Comment (borrrden) [reply src/Couchbase.Lite.Shared/API/Sync/Replicator.cs]: You could use FLValueConverter to convert it to a .NET dictionary and examine in the debugger.  Or ask Jens / Pasin what this dict is supposed to contain.

Comment (Sandychuang8) [reply src/Couchbase.Lite.Shared/API/Sync/Replicator.cs]: remove Config.PullFilter == null because it will never get to here is Config.PullFilter == null

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs [hunk 5]
```diff
@@ -116,6 +117,26 @@ public void TestEmptyPush()
             RunReplication(config, 0, 0);
         }
 
+        [Fact]
+        public void TestPushDocWithFilter()
+        {
+            using (var doc1 = new MutableDocument("doc1"))
```
Comment (borrrden) [inline src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: A better test would be to set up a database with two documents, and set up a filter that rejects one of them.  Then you can check that only one document exists in the target database (`_otherDB`).

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs [hunk 6]
```diff
@@ -161,6 +183,27 @@ public void TestPushDocContinuous()
             }
         }
 
+        [Fact]
+        public void TestPullDocWithFilter()
+        {
+            using (var doc1 = new MutableDocument("doc1")) {
```
Comment (borrrden) [inline src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: Same as the pushdocwithfilter comment for this one, except the reverse (insert into `_otherDB` and check `Db`)

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs (hunk 5 above)
Comment (Sandychuang8) [reply src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: you are right. updated.

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs (hunk 6 above)
Comment (Sandychuang8) [reply src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: updated.

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs [hunk 7]
```diff
@@ -905,6 +950,17 @@ private ReplicatorConfiguration CreateConfig(bool push, bool pull, bool continuo
         }
 #endif
 
+        private bool _replicator__filterCallback(Document doc)
+        {
+            _isFilteredCallback = true;
+            var name = doc.GetString("name");
+            if (name == "pass") {
```
Comment (borrrden) [inline src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: `return name == "pass"` ?

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs [hunk 8]
```diff
@@ -116,6 +117,27 @@ public void TestEmptyPush()
...
+                doc2.SetString("name", "pass");
+                Db.Save(doc2);
+            }
+
+            var config = CreateConfig(true, false, false);
+            config.PushFilter = _replicator__filterCallback;
+            RunReplication(config, 0, 0);
+            _isFilteredCallback.Should().BeTrue();
+            _otherDB.GetDocument("doc1").Should().BeNull("because doc1 is filtered out in the callback");
+            _otherDB.GetDocument("doc2").Count().Should().Be(1, "because doc2 is filtered in in the callback");
```
Comment (borrrden) [inline src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: This logic is slightly hard to follow.  For these purposes it is enough if the document simply is not null (other tests will test that the push operation actually transmits data correctly)

File: src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs [hunk 9]
```diff
@@ -161,6 +184,27 @@ public void TestPushDocContinuous()
...
+                doc2.SetString("name", "pass");
+                _otherDB.Save(doc2);
+            }
+
+            var config = CreateConfig(false, true, false);
+            config.PullFilter = _replicator__filterCallback;
+            RunReplication(config, 0, 0);
+            _isFilteredCallback.Should().BeTrue();
+            Db.GetDocument("doc1").Should().BeNull("because doc1 is filtered out in the callback");
+            Db.GetDocument("doc2").Count().Should().Be(1, "because doc2 is filtered in in the callback");
```
Comment (borrrden) [inline src/Couchbase.Lite.Tests.Shared/ReplicationTest.cs]: Same comment as push test.

File: src/Couchbase.Lite.Shared/API/Sync/ReplicatorConfiguration.cs [hunk 10]
```diff
@@ -69,6 +69,8 @@ public sealed class ReplicatorConfiguration
         private Authenticator _authenticator;
         private bool _continuous;
         private ReplicatorType _replicatorType = ReplicatorType.PushAndPull;
+        private Func<Document, bool> _PushFilter;
+        private Func<Document, bool> _PullValidator;
```
Comment (borrrden) [inline src/Couchbase.Lite.Shared/API/Sync/ReplicatorConfiguration.cs]: underscore lower case for consistency
